pkg/labels/selector.go

```diff
@@ -303,24 +692,19 @@ func (lsel *LabelSelector) String() (string, error) {
 // as they parse different selectors with different syntaxes.
 // The input will cause an error if it does not follow this form:
 //
-//     <selector-syntax> ::= <requirement> | <requirement> "," <selector-syntax>
-//         <requirement> ::= WHITESPACE_OPT KEY <set-restriction>
-//     <set-restriction> ::= "" | <inclusion-exclusion> <value-set>
+// <selector-syntax> ::= <requirement> | <requirement> "," <selector-syntax> ]
+// <requirement> ::= KEY [ <set-based-restriction> | <exact-match-restriction>
+// <set-based-restriction> ::= "" | <inclusion-exclusion> <value-set>
 // <inclusion-exclusion> ::= <inclusion> | <exclusion>
-//           <exclusion> ::= WHITESPACE "not" <inclusion>
-//           <inclusion> ::= WHITESPACE "in" WHITESPACE
+//           <exclusion> ::= "not" <inclusion>
+//           <inclusion> ::= "in"
 //           <value-set> ::= "(" <values> ")"
 //              <values> ::= VALUE | VALUE "," <values>
-//
-// KEY is a sequence of one or more characters that does not contain ',' or ' '
-//      [^, ]+
-// VALUE is a sequence of zero or more characters that does not contain ',', ' ' or ')'
-//      [^, )]*
-// WHITESPACE_OPT is a sequence of zero or more whitespace characters
-//      \s*
-// WHITESPACE is a sequence of one or more whitespace characters
-//      \s+
-//
+// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
+// KEY is a sequence of one or more characters that does not contain: ('=', '!', '(', ')', ',')
+// VALUE is a sequence of zero or more characters that does not contain: ('=', '!', '(', ')', ',')
```

Comment:
I'll abdicate both responsibility and authority on this - I think limited set is ok, we should reserve \ specifically for escaping in the future, but I don't have conviction on those items enough to push others that way.  I'm ok with Tim's setup.

----- Original Message -----
> >  //           <value-set> ::= "(" <values> ")"
> >  //              <values> ::= VALUE | VALUE "," <values>
> > -//
> > -// KEY is a sequence of one or more characters that does not contain ','
> > or ' '
> > -//      [^, ]+
> > -// VALUE is a sequence of zero or more characters that does not contain
> > ',', ' ' or ')'
> > -//      [^, )]*
> > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > -//      \s*
> > -// WHITESPACE is a sequence of one or more whitespace characters
> > -//      \s+
> > -//
> > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > +// KEY is a sequence of one or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> > +// VALUE is a sequence of zero or more characters that does not contain:
> > ('=', '!', '(', ')', ',')
> 
> Cross that bridge when someone complains?
> 
> On Tue, Feb 17, 2015 at 9:31 PM, Clayton Coleman <notifications@github.com>
> wrote:
> 
> > In pkg/labels/selector.go
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486#discussion_r24879872>
> > :
> >
> > >  //           <value-set> ::= "(" <values> ")"
> > >  //              <values> ::= VALUE | VALUE "," <values>
> > > -//
> > > -// KEY is a sequence of one or more characters that does not contain ','
> > > or ' '
> > > -//      [^, ]+
> > > -// VALUE is a sequence of zero or more characters that does not contain
> > > ',', ' ' or ')'
> > > -//      [^, )]*
> > > -// WHITESPACE_OPT is a sequence of zero or more whitespace characters
> > > -//      \s*
> > > -// WHITESPACE is a sequence of one or more whitespace characters
> > > -//      \s+
> > > -//
> > > +// <exact-match-restriction> ::= ["="|"=="|"!="] VALUE_NOT_EMPTY
> > > +// KEY is a sequence of one or more characters that does not contain:
> > > ('=', '!', '(', ')', ',')
> > > +// VALUE is a sequence of zero or more characters that does not contain:
> > > ('=', '!', '(', ')', ',')
> >
> > Are we going to want Unicode characters?
> > ... <#14b9b2b4f73e18a0_>
> >  On Feb 18, 2015, at 12:00 AM, Tim Hockin <notifications@github.com>
> > wrote: In pkg/labels/selector.go: > // <value-set> ::= "(" <values> ")" >
> > // <values> ::= VALUE | VALUE "," <values> > -// > -// KEY is a sequence of
> > one or more characters that does not contain ',' or ' ' > -// [^, ]+ > -//
> > VALUE is a sequence of zero or more characters that does not contain ',', '
> > ' or ')' > -// [^, )]* > -// WHITESPACE_OPT is a sequence of zero or more
> > whitespace characters > -// \s* > -// WHITESPACE is a sequence of one or
> > more whitespace characters > -// \s+ > -// > +// <exact-match-restriction>
> > ::= ["="|"=="|"!="] VALUE_NOT_EMPTY > +// KEY is a sequence of one or more
> > characters that does not contain: ('=', '!', '(', ')', ',') > +// VALUE is
> > a sequence of zero or more characters that does not contain: ('=', '!',
> > '(', ')', ',') I agree. Propose we keep label keys as qualified-name and
> > force label values as [A-Za-z0-9_-.]* - what think? ... -- Reply to this
> > email
> > directly or view it on GitHub.
> >
> > --
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24879872>
> > .
> >
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4486/files#r24882205